File: backend/app/schemas/analytics.py
from datetime import datetime

from pydantic import BaseModel


class DeliveryAnalytics(BaseModel):
    total_recipients: int
    total_deliveries: int

    pending: int
    sent: int
    delivered: int
    failed: int

    delivery_rate: float
    failure_rate: float


class ChannelAnalytics(BaseModel):
    channel: str

    total: int
    pending: int
    sent: int
    delivered: int
    failed: int

    delivery_rate: float


class CampaignAnalytics(BaseModel):
    campaign_id: str
    campaign_status: str

    total_recipients: int
    total_deliveries: int

    pending: int
    sent: int
    delivered: int
    failed: int

    delivery_rate: float
    failure_rate: float

    opens: int
    clicks: int
    responses: int
    participation: int

    open_rate: float
    click_through_rate: float
    response_rate: float
    participation_rate: float

    started_at: datetime | None
    completed_at: datetime | None

    channels: list[ChannelAnalytics]


class AnalyticsSummary(BaseModel):
    total_campaigns: int
    total_recipients: int
    total_deliveries: int

    pending: int
    sent: int
    delivered: int
    failed: int

    delivery_rate: float
    failure_rate: float

    opens: int
    clicks: int
    responses: int
    participation: int

    open_rate: float
    click_through_rate: float
    response_rate: float
    participation_rate: float

File: frontend/src/api/analytics.ts
import { apiClient } from "./client";

/* ============================================================
   CHANNEL ANALYTICS
   ============================================================ */

export interface ChannelAnalytics {
  channel: string;

  total: number;
  pending: number;
  sent: number;
  delivered: number;
  failed: number;

  delivery_rate: number;
}

/* ============================================================
   GLOBAL ANALYTICS
   GET /analytics/summary
   ============================================================ */

export interface AnalyticsSummary {
  total_campaigns: number;
  total_recipients: number;
  total_deliveries: number;

  pending: number;
  sent: number;
  delivered: number;
  failed: number;

  delivery_rate: number;
  failure_rate: number;

  opens: number;
  clicks: number;
  responses: number;
  participation: number;

  open_rate: number;
  click_through_rate: number;
  response_rate: number;
  participation_rate: number;
}

/* ============================================================
   CAMPAIGN ANALYTICS
   GET /analytics/campaign/{campaign_id}
   ============================================================ */

export interface CampaignAnalytics {
  campaign_id: string;
  campaign_status: string;

  total_recipients: number;
  total_deliveries: number;

  pending: number;
  sent: number;
  delivered: number;
  failed: number;

  delivery_rate: number;
  failure_rate: number;

  opens: number;
  clicks: number;
  responses: number;
  participation: number;

  open_rate: number;
  click_through_rate: number;
  response_rate: number;
  participation_rate: number;

  started_at: string | null;
  completed_at: string | null;

  channels: ChannelAnalytics[];
}

/* ============================================================
   GET GLOBAL ANALYTICS
   ============================================================ */

export async function fetchAnalyticsSummary(): Promise<AnalyticsSummary> {
  const response = await apiClient.get<AnalyticsSummary>(
    "/analytics/summary"
  );

  return response.data;
}

/* ============================================================
   GET CAMPAIGN ANALYTICS
   ============================================================ */

export async function fetchCampaignAnalytics(
  campaignId: string
): Promise<CampaignAnalytics> {
  const response = await apiClient.get<CampaignAnalytics>(
    `/analytics/campaign/${campaignId}`
  );

  return response.data;
}
